[DOC] Improve argument documentation for TFR multitaper

mscheltienne opened the pull request:
I've been giving a shot at TFR representation for the first time today, and it took me a while to grasp how it works. And it's still a bit of a headache for now 🤯 

With that in mind, I'd like to improve the documentation. I've been starting with multitapers, the goal of this PR is to describe better the relation between `freqs`, `n_cycles`, `time-window size`, and `time_bandwith`.
- [x] Update `tfr_array_multitaper`
- [x] Update `tfr_multitaper`
- [ ] ~~Update the tutorial `Time-frequency on simulated data (Multitaper vs. Morlet vs. Stockwell vs. Hilbert)`~~

That said I could really use an advanced course on signal processing one of those days, so please let me know if anything I wrote is wrong! 

---------

**PR Status:** for starters, I'll only update `tfr_array_multitaper`, and once the documentation is consolidated there, I will update `tfr_multitaper`, deduplicate docstrings and update the tutorial. 

Covered:
- [x] freqs
- [x] n_cycles
- [x] time-window size
- [x] time-bandwith

agramfort reviewed (commented):
looks good so far @mscheltienne 

let us know when it's ready from you end

🙏

drammock commented:
> For `tfr_multitaper`, what is the trade-off related to `time_bandwidth`? What do we lose in favor of frequency resolution and why does it have to be `>=2`?

I was recently reading about this, and found this passage from https://en.wikipedia.org/wiki/Wavelet to be clear/helpful:

> A given resolution cell's time-bandwidth product may not be exceeded with the STFT. All STFT basis elements maintain a uniform spectral and temporal support for all temporal shifts or offsets, thereby attaining an equal resolution in time for lower and higher frequencies. The resolution is purely determined by the sampling width.
> 
> In contrast, the wavelet transform's [multiresolutional](https://en.wikipedia.org/wiki/Multiresolution_analysis) properties enables large temporal supports for lower frequencies while maintaining short temporal widths for higher frequencies by the scaling properties of the wavelet transform. This property extends conventional time-frequency analysis into time-scale analysis.

In other words, time x bandwidth determines the cell "area" on the time x frequency plane (also called an "atom" sometimes). What we lose in favor of frequency resolution is temporal resolution.  IDK why it's supposed to be >=2 though.  

This figure may also be helpful: https://en.wikipedia.org/wiki/Wavelet#/media/File:Time_frequency_atom_resolution.png

mmagnuski commented:
I didn't look at this PR in detail, but some time ago I discussed a similar confusion in psd multitaper arguments on the discourse list, maybe it will be helpful: https://mne.discourse.group/t/bandwidth-of-psd/4529/2

mmagnuski commented:
Adding to @drammock description, the "Time x (Full) Bandwidth" product referred to in the docs (later just `full_TW`) is defined as `N * f` where `N` is window size (in seconds) and `f` is frequency resolution in Hz (“bandwidth”). Just as @drammock notes - you can think of it as the surface of the time x frequency windows. The number of tapers used  is `full_TW - 1`, so `full_TW` has to be at least 2 to result in a minimum of one taper.
One of the confusions present in mne docs is that psd_multitaper uses half time bandwidth product (`TW`) while tfr_multitaper uses full TW (`2 * TW`). It would be nice to make this uniform - using either full TW or TW everywhere (does not have to be in this PR of course).
Another thing that might be useful to newcomes is to expose `smoothing` (or `freq_smoothing`) argument (+- frequency smoothing, so `bandwidth / 2`, in FieldTrip it is named `tapsmofrq`), and compute the time bandwidth product out of this. For example, we have a window 1s long and the requested frequency smoothing is 4, this means that bandwidth is 8 (because smoothing is "+-", so both sides , and bandwidth is the total width). This makes full TW = 8 (`N * f = 1 * 8`) and thus 7 tapers will be used (`full_TW - 1`). But by using only the `smoothing` argument all these calculations are hidden from the user - they are faced with a control parameter that is more intuitive ("frequency smoothing" vs "full time bandwidth product"). This would require raising warinings if the requrested and effective smoothing will be different (we have only integer number of tapers that we can use, so we can't get any smoothing for given signal length) and error if requested smoothing is not possible (would result in `n_tapers < 1`).

mscheltienne commented:
Thanks for the details! 

There is one part I do not understand, how do we lose temporal resolution in favor of frequency resolution when changing the surface of the time-frequency window? The `time_bandwith` argument, `full_TW`, only impacts this surface by changing the frequency component, right?
The smaller `full_TW`, the finer the frequency resolution and the smaller the number of tapers. That's how I understand it, but what is the impact of reducing the number of tapers?

Also, currently, the `time_bandwith` argument can be defined as a float. Does this make sense as in the end `time_bandwith - 1` is floored to define the number of tapers?

mscheltienne commented:
@mmagnuski In this PR, I'll just improve the docstrings of the existing functions. Replacing the existing arguments with more intuitive ones could be done as part of @drammock PR to refactor the TFR-API, where the low-level function would retain the current set of arguments.

mmagnuski commented:
@mscheltienne Multitapering is often used when you want to increase frequency smoothing without changing window length (for example when studying gamma oscillations), so you usually want more tapers if you are doing multitapering. It also reduces the variance (noise) in your psd (in a similar way welch does, but by averaging different realisations of the same time-window signal via different tapets, not different time windows).

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3546,17 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
 
+docdict['time_bandwitdh_tfr'] = """
+time_bandwidth : float
+    Time x (Full) Bandwidth product. Should be ``≥ 2.0``.
+    Choose this along with ``n_cycles`` to set the desired frequency
+    resolution. The number of good tapers (least leakage from far away
+    frequencies) is chosen automatically based on this to
+    floor ``time_bandwidth - 1``.
+    For example, with ``freq = 20 Hz`` and ``n_cycles = 10``, we get
+    ``time = 0.5 s``. If ``time_bandwidth = 4.``, then frequency smoothing
+    is ``(4 / time) = 8 Hz``."""
```
Yes, this is still the original docstring, simply moved to the `docdict`. I first wanted to make sure I understood the trade-off made by this argument before writing about it ;)

mscheltienne commented:
>  It also reduces the variance (noise) in your psd (in a similar way welch does, but by averaging different realisations of the same time-window signal via different tapets, not different time windows).

That's what I was remembering too! Ok, now it make sense, and this is what you lose by decreasing the `full_TW` in favor of frequency resolution.

mmagnuski commented on mne/utils/docs.py:
```diff
@@ -3548,14 +3522,45 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+    smooth together the frequencies ± 2 Hz around 10 Hz, i.e. between 8 Hz and
+    12 Hz.
+
+However, the gained frequency resolution reduces the number of good tapers,
```
in multitapering you often don't want to increase the frequency resolution but increase smoothing (independent of window length), so I doub't someone would use it if their primary goal is to increase frequency resolution.

mmagnuski replied:
also, the description is very long now and better fits an example \ short tutorial, I think. I'd keep the argument description short now and add a short intro tutorial on multitapering in a next PR.

mmagnuski replied:
oh, that's a note, then the length is likely ok.

mmagnuski commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3520,48 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+    Time x (Full) Bandwidth product. Should be ``≥ 2.0``. The frequency
+    resolution achieved depends on ``n_cycles`` and ``time_bandwdith``. For a
+    given time-window length, a finer frequency resolution is achieved by
+    decreasing ``time_bandwidth`` at the cost of the number of good tapers. See
```
I wouldn't frame it this way, you use more tapers to have higher frequency smoothing (either because you want to study higher frequencies where higher frequency smoothing is beneficial or you want to reduce noice by introducing more smoothing), so if you wanted to increase frequency resolution you likely wouldn't use multitapering.

drammock reviewed (commented):
This is a big improvement!  I've left a few comments, but I think the most important one is what @mmagnuski already mentioned:  the discussion about multitaper frequency resolution has kind of the wrong emphasis. I've tried to make a suggestion here but it's not perfect so hopefully you can improve on it:

```
In multitaper spectrotemporal analysis (as with traditional fourier methods),
the temporal and spectral resolution are interrelated: longer temporal windows 
allow more precise frequency estimates; shorter temporal windows "smear"
frequency estimates while providing more precise timing information. In
MNE-Python's multitaper functions, this time-frequency trade-off is controlled
by two parameters: ``n_cycles`` and ``time_bandwidth``.

The ``n_cycles`` parameter determines the temporal window length (as a multiple
of the taper length). The ``time_bandwidth`` parameter provides the
"time-bandwidth product", which is temporal window length (in seconds) ×
frequency bandwidth (in Hz). Thus once ``n_cycles`` has been set, frequency
bandwidth is determined by :math:`\frac{time bandwidth}{time window}`, and thus
passing larger a ``time_bandwidth`` value will increase the frequency
bandwidth (thereby decreasing the frequency *resolution*).

*NB:* ``time_bandwidth`` also determines the number of "good" tapers (tapers
with minimal leakage from far-away frequencies). In MNE-Python, this is chosen
as ``floor(time_bandwidth - 1)``. This means there is another trade-off at
play, between frequency resolution and variance reduction. Striving for finer
frequency resolution (by setting ``time_bandwidth`` low) means fewer tapers
will be used, which undermines what is unique about multitaper methods — namely
their ability to improve accuracy / reduce noise in the power estimates by
using several (orthogonal) tapers; this comes at the cost of having wider
frequency bandwidths, AKA spectral smoothing.
```

mscheltienne commented on mne/time_frequency/multitaper.py:
```diff
@@ -393,10 +392,12 @@ def psd_array_multitaper(x, sfreq, fmin=0.0, fmax=np.inf, bandwidth=None,
-        or ``'power'``. If ``'power'``, the power spectral density is returned.
-        If ``output='complex'``, the complex fourier coefficients are returned
-        per taper.
+        The format of the returned ``psds`` array, ``'complex'`` or
+        ``'power'``:
+
+        * ``'power'`` : the power spectral density is returned.
+        * ``'complex'`` : the complex fourier coefficients are returned per
+          taper.
```
It's both for consistency and clarity. For this one, there are only 2 options (compared to the others with 4 or 5..), and some of them were written down as one very long sentence with each possibility comma-separated. I think it's clearer to have all possibilities listed instead.

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -807,6 +807,18 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+
+    .. note::
+        Decimation is done after convolutions and may create aliasing
+        artifacts."""
```
Yes, that will probably cause a problem in the indent count. Thanks for noticing the typo!

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3520,48 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+    decreasing ``time_bandwidth`` at the cost of the number of good tapers. See
+    notes for additional information."""
+
+docdict['time_bandwitdh_tfr_note'] = """
```
I'll keep this one open for now. I named it this way because it mostly concerns the `time_bandwidth` argument, but it can be renamed if you prefer.

I don't know why, but I have an issue with the word `width`, one time out of 2 I spell it wrong when typing; and despite some Ctrl+F to look for those mistakes, some slipped by.

mscheltienne commented on mne/time_frequency/psd.py:
```diff
@@ -287,8 +287,9 @@ def psd_multitaper(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None,
-        The bandwidth of the multi taper windowing function in Hz. The default
-        value is a window half-bandwidth of 4.
+        Half-bandwith of the multi-taper window function in Hz. For a given
+        frequency, frequencies at ``± half-bandwith`` are smoothed together.
```
Done.

mscheltienne commented on mne/time_frequency/multitaper.py:
```diff
@@ -379,12 +379,11 @@ def psd_array_multitaper(x, sfreq, fmin=0.0, fmax=np.inf, bandwidth=None,
     bandwidth : float
-        The bandwidth of the multi taper windowing function in Hz.
+        Half-bandwith of the multi-taper window function in Hz. For a given
+        frequency, frequencies at ``± half-bandwith`` are smoothed together.
```
Done, I removed the backticks.

mscheltienne commented:
Updated documentation for PSD:
- [`psd_multitaper`](https://output.circle-artifacts.com/output/job/6573c157-20d5-4073-8155-333c6d035295/artifacts/0/dev/generated/mne.time_frequency.psd_multitaper.html#mne.time_frequency.psd_multitaper) (legacy): the `bandwidth` argument docstring has been updated.
- [`psd_array_multitaper`](https://output.circle-artifacts.com/output/job/6573c157-20d5-4073-8155-333c6d035295/artifacts/0/dev/generated/mne.time_frequency.psd_array_multitaper.html#mne.time_frequency.psd_array_multitaper)
    - fmin/fmax has been replaced by `%(fmin_fmax_psd)s`.
    - the `bandwidth` and `output` argument docstrings have been updated.

Updated documentation for TFR:
- [`tfr_morlet`](https://output.circle-artifacts.com/output/job/6573c157-20d5-4073-8155-333c6d035295/artifacts/0/dev/generated/mne.time_frequency.tfr_morlet.html) and [`tfr_array_morlet`](https://output.circle-artifacts.com/output/job/6573c157-20d5-4073-8155-333c6d035295/artifacts/0/dev/generated/mne.time_frequency.tfr_array_morlet.html#mne.time_frequency.tfr_array_morlet): the `n_cycles` argument docstring has been updated and is complemented by a note.
- [`tfr_multitaper`](https://output.circle-artifacts.com/output/job/6573c157-20d5-4073-8155-333c6d035295/artifacts/0/dev/generated/mne.time_frequency.tfr_multitaper.html#mne.time_frequency.tfr_multitaper) and [`tfr_array_multitaper`](https://output.circle-artifacts.com/output/job/6573c157-20d5-4073-8155-333c6d035295/artifacts/0/dev/generated/mne.time_frequency.tfr_array_multitaper.html#mne.time_frequency.tfr_array_multitaper): the `n_cycles` and `time_bandwidth` argument docstrings have been updated and are complemented by 2 notes.

----------------

Remaining concerns:
- I used the same docstring and note for `n_cycles` in both TFRs, morlet and multitaper. But the frequency resolution/smoothing is not defined in `tfr_morlet`, and the figure in the note seems wrong for CWT. Dumb question about the morlet wavelet approach: what is the frequency resolution and its relationship with the function arguments? and the time-resolution?
- In the note about temporal window, the figure from Fieldtrip (b) shows that if the temporal window length decreases with frequency, the temporal smoothing decreases and *the frequency smoothing increases*. The note currently doesn't explain this frequency bandwidth change, e.g. if `freqs=np.arange(1, 6, 2)` and `n_cycles=2`, the temporal window length is `T=array([2., 0.7, 0.4])`, but what is the frequency bandwidth?
- The number of tapers selected by `tfr_multitaper` and `psd_multitaper` seems to defer:
    - For `psd_multitaper`, the half-bandwidth is defined in `bandwidth` which is used to create the DPSS windows here https://github.com/mne-tools/mne-python/blob/6e5ca18793c5a0d6db9c97b932bb89756c9dd17e/mne/time_frequency/multitaper.py#L440-L441

       The number of maximum tapers is defined as `int(2 * half_nbw)` https://github.com/mne-tools/mne-python/blob/6e5ca18793c5a0d6db9c97b932bb89756c9dd17e/mne/time_frequency/multitaper.py#L351-L354

    - For `tfr_multitaper`, the DPSS windows are created here https://github.com/mne-tools/mne-python/blob/6e5ca18793c5a0d6db9c97b932bb89756c9dd17e/mne/time_frequency/tfr.py#L374-L375
      and the number of max tapers provided to `dpss_windows` is https://github.com/mne-tools/mne-python/blob/6e5ca18793c5a0d6db9c97b932bb89756c9dd17e/mne/time_frequency/tfr.py#L139

  Does this difference make sense?

larsoner commented:
> Does this difference make sense?

For this question about tapers, @drammock is probably better suited to answer since IIRC he recently fixed a related bug about the number of tapers

mmagnuski commented:
> Does this difference make sense?

That is exactly the difference between psd and tfr versions of multitaper in mne that I mentioned before :)
"psd_multitaper uses half time bandwidth product (TW) while tfr_multitaper uses full TW (2 * TW)"

The difference in using `np.floor` is just cosmetic - using `int` truncates the floating point value to its integer part, so it effectively performs floor operation. So the real difference seems to be the `-1` (unless it is smuggled somewhere else in the code). But as long as `low_bias=True` asking for 5 tapers, when only 4 are valid will still give you 4. So lack of subtraction in psd code shouldn't matter. You can try it yourself:
```python
import numpy as np
from mne.time_frequency.multitaper import dpss_windows

sfreq = 250
winlen = 0.5
N = int(sfreq * winlen)
freq_bw = 10
time_bandwidth = winlen * freq_bw

n_taps_subtract = int(np.floor(time_bandwidth - 1))
n_taps_nosubtract = int(np.floor(time_bandwidth))

windows, eig = dpss_windows(N, time_bandwidth / 2., n_taps_nosubtract)

assert windows.shape[0] == n_taps_subtract
```

mscheltienne commented:
@larsoner Thanks for the explanation, I think it's clear. I'll still re-read the notes tomorrow to make sure it reads correctly. 

@mmagnuski The half-bandwidth and full-bandwidth part is very clear ;) I was worried about this -1 difference, but I'm glad to see that `dpss_windows` handles it correctly.

mscheltienne commented:
I changed the sentence order a bit in the note, I think it's better this way. Here are the updated links to the rendered doc:

Updated documentation for PSD:
- [`psd_multitaper`](https://output.circle-artifacts.com/output/job/086e27e7-2347-41bb-95f3-802805113f5c/artifacts/0/dev/generated/mne.time_frequency.psd_multitaper.html) (legacy): the `bandwidth` argument docstring has been updated.
- [`psd_array_multitaper`](https://output.circle-artifacts.com/output/job/086e27e7-2347-41bb-95f3-802805113f5c/artifacts/0/dev/generated/mne.time_frequency.psd_array_multitaper.html)
    - fmin/fmax has been replaced by `%(fmin_fmax_psd)s`.
    - the `bandwidth` and `output` argument docstrings have been updated.

Updated documentation for TFR:
- [`tfr_morlet`](https://output.circle-artifacts.com/output/job/086e27e7-2347-41bb-95f3-802805113f5c/artifacts/0/dev/generated/mne.time_frequency.tfr_morlet.html) and [`tfr_array_morlet`](https://output.circle-artifacts.com/output/job/086e27e7-2347-41bb-95f3-802805113f5c/artifacts/0/dev/generated/mne.time_frequency.tfr_array_morlet.html): the `n_cycles` argument docstring has been updated and is complemented by a note.
- [`tfr_multitaper`](https://output.circle-artifacts.com/output/job/086e27e7-2347-41bb-95f3-802805113f5c/artifacts/0/dev/generated/mne.time_frequency.tfr_multitaper.html) and [`tfr_array_multitaper`](https://output.circle-artifacts.com/output/job/086e27e7-2347-41bb-95f3-802805113f5c/artifacts/0/dev/generated/mne.time_frequency.tfr_array_multitaper.html): the `n_cycles` and `time_bandwidth` argument docstrings have been updated and are complemented by 2 notes.

drammock commented on mne/utils/docs.py:
```diff
@@ -3426,6 +3453,37 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+
+In MNE-Python, the temporal window length is defined by the arguments ``freqs``
+and ``n_cycles``, respectively defining the frequencies of interest and the
+number of cycles: :math:`T = \frac{n_{cycles}}{freqs}`
```
```suggestion
number of cycles: :math:`T = \frac{\mathtt{n\_cycles}}}{\mathtt{freqs}}`
```

drammock commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
 """
 
+docdict['time_bandwidth_tfr'] = """
+time_bandwidth : float ``≥ 2.0``
```
I'm surprised CIs don't complain about this.  In general we don't put the default on this line, but I'll admit it is concise to have it here instead of an extra sentence in the param description.

mscheltienne replied:
I was surprised as well that it did not require an exclusion in `numpydoc_xref_ignore`!

drammock commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+additionally controlled by the parameter ``time_bandwidth``.
+
+The ``n_cycles`` parameter determines the temporal window length based on the
+frequencies of interest: :math:`T = \frac{n_{cycles}}{freqs}`. The
+``time_bandwidth`` parameter defines the "time-bandwidth product", which is the
+product of the temporal window length (in seconds) and of the frequency
```
```suggestion
frequencies of interest: :math:`T = \frac{\mathtt{n\_cycles}}{\mathtt{freqs}}`.
The ``time_bandwidth`` parameter defines the "time-bandwidth product", which is
the product of the temporal window length (in seconds) and the frequency
``````

drammock commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+``time_bandwidth`` parameter defines the "time-bandwidth product", which is the
+product of the temporal window length (in seconds) and of the frequency
+bandwidth (in Hz). Thus once ``n_cycles`` has been set, frequency bandwidth is
+determined by :math:`\frac{time bandwidth}{time window}`, and thus passing a
+larger ``time_bandwidth`` value will increase the frequency bandwidth (thereby
+decreasing the frequency *resolution*).
```
```suggestion
determined by :math:`\frac{\mathrm{time bandwidth}}{\mathrm{time window}}`, and
thus passing a larger ``time_bandwidth`` value will increase the frequency
bandwidth (thereby decreasing the frequency *resolution*).
```

drammock commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+
+docdict['time_bandwidth_tfr_notes'] = r"""
+In MNE-Python's multitaper functions, the time-frequency trade-off is
+additionally controlled by the parameter ``time_bandwidth``.
+
```
```suggestion
additionally affected by the parameter ``time_bandwidth``.
```

drammock commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+larger ``time_bandwidth`` value will increase the frequency bandwidth (thereby
+decreasing the frequency *resolution*).
+
+*NB:* ``time_bandwidth`` also determines the number of "good" tapers (tapers
+with minimal leakage from far-away frequencies). In MNE-Python, this is chosen
+as ``floor(time_bandwidth - 1)``. This means there is another trade-off at
+play, between frequency resolution and variance reduction. Striving for finer
+frequency resolution (by setting ``time_bandwidth`` low) means fewer tapers
+will be used, which undermines what is unique about multitaper methods — namely
+their ability to improve accuracy / reduce noise in the power estimates by
+using several (orthogonal) tapers; this comes at the cost of having wider
+frequency bandwidths, AKA spectral smoothing.
```
Some slight rewording for clarity

```suggestion
Note that ``time_bandwidth`` also determines the number of tapers used.
MNE-Python uses only "good" tapers (tapers with minimal leakage from far-away
frequencies); the number of good tapers is ``floor(time_bandwidth - 1)``. This
means there is another trade-off at play, between frequency resolution and the
variance reduction that multitaper analysis provides. Striving for finer
frequency resolution (by setting ``time_bandwidth`` low) means fewer tapers
will be used, which undermines what is unique about multitaper methods — namely
their ability to improve accuracy / reduce noise in the power estimates by
using several (orthogonal) tapers. This variance reduction comes at the cost of
having wider frequency bandwidths.
```

drammock reviewed (approved):
This is looking really good.  I've made a few suggestions of minor wording improvements, and some LaTeX tweaks.  Thanks for taking this on, it's a big improvement.

mmagnuski commented:
Thanks for working on this @mscheltienne, I'll take a look later today!

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -3426,6 +3453,37 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+Either the temporal window has a fixed length independent of frequency, or the
+temporal window decreases in length with increased frequency.
+
+.. image:: https://www.fieldtriptoolbox.org/assets/img/tutorial/timefrequencyanalysis/figure1.png
```
Completely agree.

mscheltienne commented:
I think there is a bracket typo in some of the LaTex syntax, I'll check if the doc renders correctly.

drammock commented:
> I think there is a bracket typo in some of the LaTex syntax, I'll check if the doc renders correctly.

I wouldn't be surprised :)  I didn't do a local test sphinx build 🙈

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -3426,6 +3453,37 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+
+In MNE-Python, the temporal window length is defined by the arguments ``freqs``
+and ``n_cycles``, respectively defining the frequencies of interest and the
+number of cycles: :math:`T = \frac{\mathtt{n\_cycles}}}{\mathtt{freqs}}`
```
```suggestion
number of cycles: :math:`T = \frac{\mathtt{n\_cycles}}{\mathtt{freqs}}`
```

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+The ``time_bandwidth`` parameter defines the "time-bandwidth product", which is
+the product of the temporal window length (in seconds) and the frequency
+bandwidth (in Hz). Thus once ``n_cycles`` has been set, frequency bandwidth is
+determined by :math:`\frac{\mathrm{time bandwidth}}{\mathrm{time window}}`, and
```
@drammock The spaces don't render in this one, do you know how to fix it? Should we escape the spaces?

![Screenshot 2022-11-01 at 17 14 10](https://user-images.githubusercontent.com/73893616/199282288-63d82cf9-733a-4596-b6d2-c67f2a7990df.png)

drammock replied:
oops, should be `time~bandwidth` and `time~window`

mscheltienne commented on mne/utils/docs.py:
```diff
@@ -3495,6 +3553,50 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+The ``time_bandwidth`` parameter defines the "time-bandwidth product", which is
+the product of the temporal window length (in seconds) and the frequency
+bandwidth (in Hz). Thus once ``n_cycles`` has been set, frequency bandwidth is
+determined by :math:`\frac{\mathrm{time bandwidth}}{\mathrm{time window}}`, and
```
```suggestion
determined by :math:`\frac{\mathrm{time~bandwidth}}{\mathrm{time~window}}`, and
```

mscheltienne commented:
The LaTex part renders correctly now ;)

drammock commented:
I'm happy 👍🏻 @mmagnuski I'll let you merge if you're satisfied. Here's a quick link to the CircleCI docs build for [tfr_multitaper](https://output.circle-artifacts.com/output/job/366f7ae9-baa0-4bdb-ae02-78986207fd7d/artifacts/0/dev/generated/mne.time_frequency.tfr_multitaper.html) and [tfr_morlet](https://output.circle-artifacts.com/output/job/366f7ae9-baa0-4bdb-ae02-78986207fd7d/artifacts/0/dev/generated/mne.time_frequency.tfr_morlet.html)

note the CIs won't go green because of `skip` annotations; Circle was green so OK to merge without waiting for other CIs.

mmagnuski commented:
Great work @mscheltienne! The only remaining thing for me is changing the focus in the multitaper-specific part of the notes a bit:
I would not discuss `time_bandwidth` as related to time-frequency trade-off. The resolution in time is unaffected here - the `n_cycles` stays as it was defined. Changing the `time_bandwidth` allows to introduce **additional** frequency smoothing. So I would just focus on the fact that in multitapering we can use `time_bandwidth` to increase frequency smoothing independent of window length to decrease noise in our psd estimate (as you nicely discuss in the n_tapers paragraph). It would be useful to also mention that it is frequently done when studying higher frequencies such as gamma (another application is in sleep research - especially when plotting the whole-night spectrogram: for a single participant there is only one trial then so reducing noise is useful).
Also when discussing number of tapers instead of "Note that time_bandwidth also determines the number of tapers used." I would prefer to say that increased frequency smoothing in multitapering is realised by averaging psds obtained from different tapers, so to get higher smoothing one needs to use more tapers, and then explain the relationship as you do now.

## more details
The additional frequency smoothing in multitapering may be desirable for at least two reasons: decreasing the noise in the spectrum and increasing the frequency overlap between oscillatory events (across trials for example when performing averaging) when you are interested in phenomena that occur in a relatively wide frequency window (like gamma oscillations).
There is a convention running in the FieldTrip/Donders group to perform multitapering (meaning > 1 taper) for frequencies above 30 Hz. You can find this recommendation in the FLUX pipeline paper from Ole Jensen lab ("FLUX: A pipeline for MEG analysis"). They reccomend 0.25 s window with five tapers (+- 12 Hz spectral smoothing). They don't give detailed explanation for why the smoothing is beneficial (apart from that oscillations "are more broadband in the gamma range"), but as I write earlier I think it is to both decrease the nosie and increase single trial gamma events overlap when averaging. But both these reasons (yet another is increasing overlap in frequency between participants for statistics) relate to different aspects of noise so they can be discussed this way in the note.

mscheltienne commented:
`tfr_multitaper`: https://output.circle-artifacts.com/output/job/d2312936-b67d-4e0f-8db3-1c0eecc298d4/artifacts/0/dev/generated/mne.time_frequency.tfr_multitaper.html

@mmagnuski You mentioned several times that multitapering is interesting when applied to a wide frequency window, and especially to high frequencies. Is it however a "bad" method to apply on lower frequency bands, delta, theta, alpha?
If so, then I think it should be clearly mentioned at the beginning of the docstring.

Also, the increased smoothing when studying high-frequencies comes from the bandwidth. Thus, if we leave a bit TFRs and return to PSDs, how would you compute a power estimate on short recordings in the lower frequencies? I would assume a multitaper-psd with only a few tapers (e.g. 3) and a finer frequency resolution would still provide a better estimate than a welch-psd with only a few windows.

mmagnuski commented:
Great, thanks for the changes @mscheltienne! I'll merge the PR in a few minutes.
Answering your questions:

> Is it however a "bad" method to apply on lower frequency bands, delta, theta, alpha?
If so, then I think it should be clearly mentioned at the beginning of the docstring.

No, often there is just no need to use multitapering for these lower frequencies (the signal to noise i usually high, peaks easily discernible in the spectrum). But I wouldn't explicitly discourge using multitapering for lower frequencies. I can image it might be beneficial when studying single-trial TFR/spectra. 

> how would you compute a power estimate on short recordings in the lower frequencies? I would assume a multitaper-psd with only a few tapers (e.g. 3) and a finer frequency resolution would still provide a better estimate than a welch-psd with only a few windows.

Yes, for short segments, multitapering is a good alternative to Welch.

mmagnuski commented:
I don't think I can merge this without all required tests being green. All I was able to do was enable auto-merge.
@mscheltienne - could you push just one small change but without skip commands so that all test are run? Unless there is an easier way to merge this. :)

mmagnuski commented:
BTW @larsoner how were you able to merge this?

larsoner commented:
I think we configured this repo so that people with repo admin-level permissions can override required CI checks, basically for situations like this or when CIs are misbehaving
